src/lib/token_tree.js

```diff
@@ -0,0 +1,92 @@
+export default class TokenTree {
+  constructor() {
+    this.rootNode = { root: true, children: [] };
+    this.stack = [ this.rootNode ];
+  }
+
+  get top() {
+    return this.stack[this.stack.length - 1];
+  }
+
+  add(node) {
+    this.top.children.push(node);
+  }
+
+  addKeyword(text, kind) {
+    if (text === "") { return; }
+
+    this.top.children.push({
+      kind,
+      children: [ text ]
+    });
+  }
```

Comment:
It looks like it is essentially `openNode` + `addText` + `closeNode`. So do we need this method? Its name makes me think that a "keyword" is something special in the tree, but it is not.

Reply:
>  Its name makes me think that a "keyword" is something special in the tree, but it is not.

Here tree is just an implementation of an abstract interface.  Javascript doesn't have interfaces that I'm aware of... I'm defining the simplest API that is useful for the parser and then NodeTree (or whatever other object) is simply fulfilling that interface.

I found addKeyword much more useful than repeating the pattern you suggest over and over. 

I suppose you could argue that technically the emitter object should be separate from the implementation of the tree object... but I'm not sure it's worth it here since the objects are already so dead simple.

Reply:
> I found addKeyword much more useful than repeating the pattern you suggest over and over.

It is called from exactly one place, actually. To my taste it introduces more complexity. At least, it could be implemented in terms of already defined methods (openNode, addText, closeNode). Otherwise it takes some time to understand that it's not something new but simply a shortcut.

Reply:
> It is called from exactly one place, actually.

Ok that takes the wind out of my "over and over" sails, lol.  But I still think that doesn't mean it isn't an helpful abstraction.  It's my "separation of concerns" voice saying the parser doesn't need to know that much about the structure of the tree... hence treating it as an emitter object and just telling it "emit(keyword)", etc.

> At least, it could be implemented in terms of already defined methods

Now that's a useful observation for sure, although I worry it'd be slower - but since we have no real large benchmarks and premature optimization and all I'll refactor it to build on the existing blocks.

Reply:
Interestingly, `npm test` runs on my laptop 8s, and 7s with this PR applied. So in fact your refactoring speeds things up. Cool.

Reply:
> Interestingly, npm test runs on my laptop 8s, and 7s with this PR applied. So in fact your refactoring speeds things up. Cool.

Oh - that's interesting?  Try measuring the memory usage though.  I only tried it once but it was like double or so.

Reply:
With zsh `time`:

parse tree:

```
avg shared (code):         0 KB
avg unshared (data/stack): 0 KB
total (sum):               0 KB
max memory:                191216 MB
page faults from disk:     0
other page faults:         64816
```

master:

```
avg shared (code):         0 KB
avg unshared (data/stack): 0 KB
total (sum):               0 KB
max memory:                227424 MB
page faults from disk:     0
other page faults:         73349
```

So now it seems parse tree is the clear winner in all regards, hmmmmm...

Reply:
I'm not sure WHY it's faster or more memory performant though... I'd have bet the opposite on both counts.

Reply:
I got similar figures... but I'm not sure if `time` counts child processes?
I'm running `time -v npm test` and `ps` show two processes: `node` and `npm`.

Reply:
I run the tests directly usually, so I'm not even invoking `npm` at all. :)

Reply:
@marcoscaceres Any idea why we'd see these results?  I mean we're still building the huge output string, as before - just now we're ALSO first building a huge tree of nodes and then walking it to generate the output string.

Reply:
Ok, I'm not completely crazy.  When running a single test of a 1.2mb JS file (our own built source) I get ~88mb with master and ~93mb with parse tree.  So "it depends"... but with such a small increase (and even decrease for smaller workloads) I don't think I have a problem with the parse tree being the new default.

Reply:
Agreed.
A profiling tool should give insights, but that's another story. I think we're done with it by now.